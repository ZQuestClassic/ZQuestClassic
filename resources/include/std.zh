// std.zh
// 28th December, 2020 for ZC 2.55 Alpha 86
// Standard header for ZScript.
// Please don't edit this file. Instead, create a new header file for your
// quests' scripts,  add constants and procedures to that file,
// and #include it separately.

//Options
#option BINARY_32BIT off
	//Would break a number of things if set on
#option SHORT_CIRCUIT on
	//Should work for everything in std_zh, and just speed it up

#includepath "include"

//Version Metadata
#include "std_zh/std_meta.zh"

//User Configuration Files
#include "std_zh/std_user_defs.zh" //Typedefs: Always first.
#include "std_zh/std.cfg"

//Main Files
#include "std_zh/limits.zh"
#include "std_zh/std_constants.zh"
#include "std_zh/ffrules.zh"
#include "std_zh/std_functions.zh"
#include "std_zh/math.zh"
#include "std_zh/weapon.zh"
#include "std_zh/std_sideview.zh" //Sideview and platforming

//ZScript string handling functions, akin to those found in C
//#include "string.zh"
#include "std_zh/string/string_constants.zh" //Now #includeed from here. This prevents conflicts for those people
#include "std_zh/string/string_functions.zh" //who would #include string.zh from a secondary source, as that file is now 
			     //empty, and the string.zh components load from here. 

//Keyboard handling for 2.54 and above
#include "std_zh/std_keyboard.zh"

//Locale
#include "std_zh/locale/std_EnglishIntl.zh" //International English Localisation

//Legacy Wrappers
#include "std_zh/std_legacy.zh"