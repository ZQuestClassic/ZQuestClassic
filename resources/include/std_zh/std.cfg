//Settings

const int STD_FORCE_OLD_DISTANCE = 1;	//If enabled, Distance() will use Sqrt() instead of SafeSqrt().
					//! If you use this, disable the quest rule 'Log Script Errors to Allegro.log
					//! before distributing your quest!
					
const int STD_DISTANCE_USE_SAFESQRT = 0; 	//If enebaled, Distance will use SafeSqrt(). Otherwise, it will return a factor, 
					
const int STD_EMULATE_ENGINE_COLLISION = 0; // If enabled, Collision hitboxs will more precisely emulate the
					    // internal collision. This differs to how scripted collision
					    // has always worked.
					    

const int STD_CENTERXY_USES_HITBOXES = 0; // If enabled, CenterX() and CenterY() will use object hitboxes
					  // and offsets by default, rather than sprite sizes when calculating
					  // the centre of objects.
					  

const int STD_NEVER_USE_SAFESQRT = 1; //Enable to force any calls to SafeSqrt() to use Sqrt() instead.

const int MAX_HIGHEST_LEVEL_ITEM_CHECK = 255; //The highest item ID checked by GetHighestLevelItem(). 

const int REVISED_ROTATEPOINT_X_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointX()
const int REVISED_ROTATEPOINT_Y_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointY()


const int MAX_NOACTION_INPUT = 14; //Checks A, B, L, R, U/D/L/R, Ex1, Ex2, Ex3, Ex4
	//Use CB_MAX to force the analogue stick false. 
const int NOACTION_SKIP_MAP = 0; //If enabled, NoAction will not disable pressing the Map button
const int NOACTION_SKIP_START = 0; //If enabled, NoAction will not disable pressing the Start button

const int OLD_ONSIDEPLATFORM = 0; //Change to 1 to use old OnSidePlatform calls.

const int OLD_RAND_FUNCS = 0; //Change to 1 to use old Rand() logic
