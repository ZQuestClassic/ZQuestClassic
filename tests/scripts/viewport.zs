import "std.zh"

global script Global
{
    void run()
    {
        Viewport->X = 100;
    }
}
