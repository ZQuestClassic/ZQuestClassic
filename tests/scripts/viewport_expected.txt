Compiling 'viewport.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

PUSHARGSV 0,0.0004; void run() Body Start
SETR d4,SP
STOREDV 20,2
STOREDV 5,0
SETV d2,2; while(true) #8 Body [Opt:AlwaysOn] Start
PUSHR d2
POP d2
SETR VIEWPORT_MODE,d2
SETR d2,LINKX
PUSHR d2
SETR d2,VIEWPORT_WIDTH
DIVV d2,2
POP d3
SUBR d3,d2
SETR d2,d3
PUSHR d2
LOADD d2,2
PUSHR d2
LOADD d2,1; InlineFunc[int Cos(int)] Params Start
SETR d3,d2; InlineFunc[int Cos(int)] Params End
COSR d2,d3; InlineFunc[int Cos(int)] Body
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
PUSHR d2
POP d2
SETR VIEWPORT_X,d2
SETR d2,LINKY
PUSHR d2
SETR d2,VIEWPORT_HEIGHT
DIVV d2,2
POP d3
SUBR d3,d2
SETR d2,d3
PUSHR d2
LOADD d2,2
PUSHR d2
LOADD d2,1; InlineFunc[int Sin(int)] Params Start
SETR d3,d2; InlineFunc[int Sin(int)] Params End
SINR d2,d3; InlineFunc[int Sin(int)] Body
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
PUSHR d2
POP d2
SETR VIEWPORT_Y,d2
LOADD d2,1
PUSHR d2
LOADD d2,0
POP d3
ADDR d2,d3
STORED d2,1; if() #189 Test
COMPAREV d2,-360
SETCMP d2,I<
COMPAREV d2,0
GOTOCMP 63,==; Test 'if'
LOADD d2,1; if() #189 Body Start
ADDV d2,360
STORED d2,1
GOTO 71; if() #189 Body End
LOADD d2,1; if() #558 Test AND if() #189 Else Start
COMPAREV d2,360
SETCMP d2,I>
COMPAREV d2,0
GOTOCMP 71,==; Test 'if'
LOADD d2,1; if() #558 Body Start
SUBV d2,360
STORED d2,1; if() #558 Body End
WAITFRAME; if() #189 Else End AND InlineFunc[void Waitframe()] Body AND while(true) #8 Body [Opt:AlwaysOn] End
GOTO 5; while() #8 End AND [Opt:NeverRet] AND void run() Body End


ALLOCATEGMEMV d2,4096; void run() Body Start
SETR gd1,d2
QUIT; void run() Body End